diff -Naur "a/ecm_classifier.c" b/ecm_classifier.c
--- "a/ecm_classifier.c"	2024-11-06 13:33:31.000000000 +0100
+++ b/ecm_classifier.c	2026-08-05 16:57:14.807991793 +0200
@@ -17,6 +17,8 @@
  **************************************************************************
  */
 
+
+#include <linux/printk.h>
 #include <linux/version.h>
 #include <linux/types.h>
 #include <net/ip.h>
@@ -91,6 +93,9 @@
 	case ECM_CLASSIFIER_TYPE_PCC: {
 		struct ecm_classifier_pcc_instance *pcci;
 
+		/* DEBUG */
+		// printk(KERN_INFO "MARK: Attempting to assign classifier\n");
+
 		pcci = ecm_classifier_pcc_instance_alloc(ci);
 		if (!pcci) {
 			DEBUG_TRACE("%px: Failed to create Parental Controls classifier\n", ci);
@@ -175,6 +180,9 @@
 	case ECM_CLASSIFIER_TYPE_MARK: {
 		struct ecm_classifier_mark_instance *ecmi;
 
+		/* DEBUG */
+		// printk(KERN_INFO "MARK: assign_classifier() called for MARK type\n");
+
 		ecmi = ecm_classifier_mark_instance_alloc(ci);
 		if (!ecmi) {
 			DEBUG_TRACE("%px: Failed to create mark classifier\n", ci);
@@ -230,6 +238,8 @@
  */
 bool ecm_classifier_reclassify(struct ecm_db_connection_instance *ci, int assignment_count, struct ecm_classifier_instance *assignments[])
 {
+	/* DEBUG printk */
+	// printk(KERN_INFO "CLASSIFIER: reclassify() called for serial %u\n", ecm_db_connection_serial_get(ci));
 	ecm_classifier_type_t classifier_type;
 	int i;
 	bool full_reclassification = true;
@@ -238,6 +248,9 @@
 	 * assignment_count will always be <= the number of classifier types available
 	 */
 	for (i = 0, classifier_type = ECM_CLASSIFIER_TYPE_DEFAULT; i < assignment_count; ++i, ++classifier_type) {
+
+
+
 		ecm_classifier_type_t aci_type;
 		struct ecm_classifier_instance *aci;
 
@@ -278,6 +291,8 @@
 	 */
 	for (; classifier_type < ECM_CLASSIFIER_TYPES; ++classifier_type) {
 		struct ecm_classifier_instance *naci;
+		/* DEBUG printk */
+		// printk(KERN_INFO "CLASSIFIER: Trying to assign type %d\n", classifier_type);
 		DEBUG_TRACE("%px: Instantiate missing type: %d\n", ci, classifier_type);
 
 		naci = ecm_classifier_assign_classifier(ci, classifier_type);
diff -Naur "a/ecm_classifier_default.c" b/ecm_classifier_default.c
--- "a/ecm_classifier_default.c"	2024-11-06 13:33:31.000000000 +0100
+++ b/ecm_classifier_default.c	2026-08-05 16:57:14.811991793 +0200
@@ -1,5 +1,4 @@
 /*
- **************************************************************************
  * Copyright (c) 2014-2016, 2020-2021, The Linux Foundation. All rights reserved.
  * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
  *
@@ -14,9 +13,9 @@
  * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
  * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
- **************************************************************************
  */
 
+#include <linux/printk.h>
 #include <linux/version.h>
 #include <linux/types.h>
 #include <linux/ip.h>
@@ -33,7 +32,7 @@
 #include <net/ip.h>
 #include <net/tcp.h>
 #include <asm/unaligned.h>
-#include <asm/uaccess.h>	/* for put_user */
+#include <asm/uaccess.h>
 #include <net/ipv6.h>
 #include <linux/inet.h>
 #include <linux/in.h>
@@ -48,7 +47,6 @@
 #include <net/netfilter/nf_conntrack_core.h>
 #include <net/netfilter/ipv4/nf_conntrack_ipv4.h>
 #include <net/netfilter/ipv4/nf_defrag_ipv4.h>
-
 /*
  * Debug output levels
  * 0 = OFF
@@ -72,69 +70,47 @@
 #include "ecm_classifier_default.h"
 #include "ecm_front_end_common.h"
 
-/*
- * Magic numbers
- */
 #define ECM_CLASSIFIER_DEFAULT_INTERNAL_INSTANCE_MAGIC 0x8761
 #define ECM_CLASSIFIER_DEFAULT_STATE_FILE_INSTANCE_MAGIC 0x3321
 
-/*
- * struct ecm_classifier_default_internal_instance
- * 	State to allow tracking of dynamic priority for a connection
- */
 struct ecm_classifier_default_internal_instance {
-	struct ecm_classifier_default_instance base;		/* Base type */
-
-	uint32_t ci_serial;					/* RO: Serial of the connection */
-	int protocol;						/* RO: Protocol of the connection */
-
+	struct ecm_classifier_default_instance base;
+	uint32_t ci_serial;
+	int protocol;
 	struct ecm_classifier_process_response process_response;
-								/* Last process response computed */
-
-	ecm_db_timer_group_t timer_group;			/* The timer group the connection should be in based on state */
-
-	ecm_tracker_sender_type_t ingress_sender;		/* RO: Which sender is sending ingress data */
-	ecm_tracker_sender_type_t egress_sender;		/* RO: Which sender is sending egress data */
-
-	struct ecm_tracker_instance *ti;			/* RO: Tracker used for state and timer group checking. Pointer will not change so safe to access outside of lock. */
-	bool packet_seen[ECM_CONN_DIR_MAX];                     /* Per-direction packet seen flag */
-	int refs;						/* Integer to trap we never go negative */
-#if (DEBUG_LEVEL > 0)
+	ecm_db_timer_group_t timer_group;
+	ecm_tracker_sender_type_t ingress_sender;
+	ecm_tracker_sender_type_t egress_sender;
+	struct ecm_tracker_instance *ti;
+	bool packet_seen[ECM_CONN_DIR_MAX];
+	int refs;
+	bool classified;
+	bool decelerated;
+	#if (DEBUG_LEVEL > 0)
 	uint16_t magic;
-#endif
+	#endif
 };
 
-static  DEFINE_SPINLOCK(ecm_classifier_default_lock);			/* Concurrency control SMP access */
-static int ecm_classifier_default_count = 0;			/* Tracks number of instances allocated */
+static DEFINE_SPINLOCK(ecm_classifier_default_lock);
+static int ecm_classifier_default_count = 0;
 
-/*
- * Operational control
- */
 static ecm_classifier_acceleration_mode_t ecm_classifier_default_accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL;
-								/* Cause connections whose hosts are both on-link to be accelerated */
-static int ecm_classifier_default_enabled = 1;		/* When disabled the qos algorithm will not be applied to skb's */
-
-/*
- * Management thread control
- */
-static bool ecm_classifier_default_terminate_pending = false;	/* True when the user wants us to terminate */
+static int ecm_classifier_default_enabled = 1;
+static bool ecm_classifier_default_terminate_pending = false;
 
-/*
- * Character device stuff - used to communicate status back to user space
- */
 #define ECM_CLASSIFIER_DEFAULT_STATE_FILE_BUFFER_SIZE 1024
 struct ecm_classifier_default_state_file_instance {
 	struct ecm_classifier_default_internal_instance *cdii;
 	bool doc_start_written;
 	bool doc_end_written;
-	char msg_buffer[ECM_CLASSIFIER_DEFAULT_STATE_FILE_BUFFER_SIZE];	/* Used to hold the current state message being output */
-	char *msgp;							/* Points into the msg buffer as we output it piece by piece */
-	int msg_len;							/* Length of the buffer still to be written out */
-#if (DEBUG_LEVEL > 0)
+	char msg_buffer[ECM_CLASSIFIER_DEFAULT_STATE_FILE_BUFFER_SIZE];
+	char *msgp;
+	int msg_len;
+	#if (DEBUG_LEVEL > 0)
 	uint16_t magic;
-#endif
+	#endif
 };
-static struct dentry *ecm_classifier_default_dentry;		/* Debugfs dentry object */
+static struct dentry *ecm_classifier_default_dentry;
 
 /*
  * _ecm_classifier_default_ref()
@@ -182,22 +158,13 @@
 		return refs;
 	}
 
-	/*
-	 * Object to be destroyed
-	 */
 	ecm_classifier_default_count--;
 	DEBUG_ASSERT(ecm_classifier_default_count >= 0, "%px: ecm_classifier_default_count wrap\n", cdii);
 
 	spin_unlock_bh(&ecm_classifier_default_lock);
 
-	/*
-	 * Release our tracker
-	 */
 	cdii->ti->deref(cdii->ti);
 
-	/*
-	 * Final
-	 */
 	DEBUG_INFO("%px: Final default classifier instance\n", cdii);
 	kfree(cdii);
 
@@ -205,238 +172,129 @@
 }
 
 /*
- * ecm_classifier_default_ready_for_accel()
- *	Checks if the connection is ready for the acceleration.
- *
- * This function is called, if the acceleration delay feature is enabled.
- */
-static bool ecm_classifier_default_ready_for_accel(
-			struct ecm_classifier_default_internal_instance *cdii,
-			ecm_tracker_sender_type_t sender)
-{
-	uint64_t slow_pkts;
-	struct ecm_db_connection_instance *ci;
-	struct ecm_front_end_connection_instance *feci;
-
-	/*
-	 * We delay the acceleration for TCP and UDP protocols only.
-	 * Non-ported protocols mostly the outer connections of the flows.
-	 * Acceleration delay is required to inspect the packets and
-	 * the outer connections are generally do not need inspection.
-	 */
-	if (cdii->protocol != IPPROTO_TCP && cdii->protocol != IPPROTO_UDP) {
-		DEBUG_TRACE("%px: Accel delay is not enabled for protocol: %d\n", cdii, cdii->protocol);
-		return true;
-	}
-
-	/*
-	 * Delay forever until seeing the reply packet.
-	 */
-	if (ecm_classifier_accel_delay_pkts == 1) {
-		DEBUG_INFO("%px: Checking both direction traffic\n", cdii);
-
-		/*
-		 * Set the flow and return direction packet seen flags.
-		 */
-		if (sender == ECM_TRACKER_SENDER_TYPE_SRC) {
-			cdii->packet_seen[ECM_CONN_DIR_FLOW] = true;
-		} else {
-			cdii->packet_seen[ECM_CONN_DIR_RETURN] = true;
-		}
-
-		/*
-		 * Check if both direction traffic is seen.
-		 */
-		return cdii->packet_seen[ECM_CONN_DIR_FLOW] && cdii->packet_seen[ECM_CONN_DIR_RETURN];
-	}
-
-	/*
-	 * Delay the acceleration until we see <N> number of packets.
-	 * ecm_classifier_accel_delay_pkts = <N>
-	 */
-	ci = ecm_db_connection_serial_find_and_ref(cdii->ci_serial);
-	if (!ci) {
-		DEBUG_TRACE("%px: No ci found for %u\n", cdii, cdii->ci_serial);
-		return false;
-	}
-
-	/*
-	 * Get the packet count we have seen in the slow path so far.
-	 */
-	feci = ecm_db_connection_front_end_get_and_ref(ci);
-	slow_pkts = ecm_front_end_get_slow_packet_count(feci);
-	ecm_front_end_connection_deref(feci);
-	ecm_db_connection_deref(ci);
-
-	/*
-	 * Check if we have seen slow path packets as the predefined count.
-	 */
-	if (slow_pkts < ecm_classifier_accel_delay_pkts) {
-		DEBUG_TRACE("%px: delay the acceleration: slow packets: %llu default delay packet count: %d\n",
-			    cdii, slow_pkts, ecm_classifier_accel_delay_pkts);
-
-		/*
-		 * We haven't reached the slow path packet limit.
-		 * We can wait more to accelerate the connection.
-		 */
-		return false;
-	}
-
-	/*
-	 * We waited enough time for the acceleration, we can allow it now.
-	 */
-	DEBUG_INFO("%px: Let the flow accel, waited enough packet\n", cdii);
-	return true;
-}
-
-/*
  * ecm_classifier_default_process()
  *	Process the flow for acceleration decision.
  */
 static void ecm_classifier_default_process(struct ecm_classifier_instance *aci, ecm_tracker_sender_type_t sender,
-					   struct ecm_tracker_ip_header *ip_hdr, struct sk_buff *skb,
-					   struct ecm_classifier_process_response *process_response)
+										   struct ecm_tracker_ip_header *ip_hdr, struct sk_buff *skb,
+										   struct ecm_classifier_process_response *process_response)
 {
-	struct ecm_tracker_instance *ti;
-	ecm_tracker_sender_state_t from_state;
-	ecm_tracker_sender_state_t to_state;
-	ecm_tracker_connection_state_t prevailing_state;
-	ecm_db_timer_group_t tg;
 	struct ecm_classifier_default_internal_instance *cdii = (struct ecm_classifier_default_internal_instance *)aci;
-	struct nf_conn *ct;
-	enum ip_conntrack_info ctinfo;
-
-	DEBUG_CHECK_MAGIC(cdii, ECM_CLASSIFIER_DEFAULT_INTERNAL_INSTANCE_MAGIC, "%px: invalid state magic\n", cdii);
+	struct ecm_db_connection_instance *ci = NULL;
+	struct ecm_front_end_connection_instance *feci = NULL;
+	struct ecm_classifier_instance *mark_classi = NULL;
+	struct ecm_classifier_process_response mark_pr;
+	bool mark_denies = false;
+	bool mark_present = false;
+	bool is_syn = false;
+
+	/* DEBUG: Verificar si es un SYN (TCP) */
+	if (skb && skb->protocol == ETH_P_IP) {
+		struct iphdr *ip = (struct iphdr *)skb->data;
+		if (ip->protocol == IPPROTO_TCP) {
+			struct tcphdr *tcp = (struct tcphdr *)(skb->data + (ip->ihl * 4));
+			if (tcp->syn && !tcp->ack) {
+				is_syn = true;
+				// printk(KERN_INFO "DEFAULT: SYN packet detected for serial %u\n", cdii->ci_serial);
+			}
+		}
+	}
 
-	/*
-	 * Get qos result and accel mode
-	 * Default classifier is rarely disabled.
-	 */
-	if (unlikely(!ecm_classifier_default_enabled)) {
-		/*
-		 * Still relevant but have no actions that need processing
-		 */
+	/* DEBUG: SI NO ES SYN, NO PROCESAR (solo el SYN decide el offload) */
+	if (!is_syn) {
 		spin_lock_bh(&ecm_classifier_default_lock);
-		cdii->process_response.process_actions = 0;
-		*process_response = cdii->process_response;
+		if (cdii->classified) {
+			*process_response = cdii->process_response;
+			spin_unlock_bh(&ecm_classifier_default_lock);
+			return;
+		}
 		spin_unlock_bh(&ecm_classifier_default_lock);
-		return;
 	}
 
-	/*
-	 * Update connection state
-	 * Compute the timer group this connection should be in.
-	 * For this we need the tracker and the state to be updated.
-	 * NOTE: Tracker does not need to be ref'd it will exist for as long as this default classifier instance does
-	 * which is at least for the duration of this call.
-	 */
-	ti = cdii->ti;
-	ti->state_update(ti, sender, ip_hdr, skb);
-	ti->state_get(ti, &from_state, &to_state, &prevailing_state, &tg);
+	// printk(KERN_INFO "DEFAULT: process() called for serial %u\n", cdii->ci_serial);
+
+	/* Verifica si ya está clasificada  */
 	spin_lock_bh(&ecm_classifier_default_lock);
-	if (unlikely(cdii->timer_group != tg)) {
-		/*
-		 * Timer group has changed
-		 */
-		cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_TIMER_GROUP;
-		cdii->process_response.timer_group = tg;
-
-		/*
-		 * Record for future change comparisons
-		 */
-		DEBUG_INFO("%px: timer group changed from %d to %d\n", cdii, cdii->timer_group, tg);
-		cdii->timer_group = tg;
-	}
-
-	/*
-	 * Don't care?
-	 */
-	if (ecm_classifier_default_accel_mode == ECM_CLASSIFIER_ACCELERATION_MODE_DONT_CARE) {
-		cdii->process_response.process_actions &= ~ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+	if (cdii->classified) {
 		*process_response = cdii->process_response;
 		spin_unlock_bh(&ecm_classifier_default_lock);
+		// printk(KERN_INFO "DEFAULT: Already classified, returning previous decision\n");
 		return;
 	}
 	spin_unlock_bh(&ecm_classifier_default_lock);
 
-	/*
-	 * Handle non-TCP case
-	 */
-	if (cdii->protocol != IPPROTO_TCP) {
-		if (unlikely(prevailing_state != ECM_TRACKER_CONNECTION_STATE_ESTABLISHED)) {
-			DEBUG_INFO("%px: Protocol: %d is non-TCP, prevailing_state: %d\n", cdii, cdii->protocol, prevailing_state);
-			spin_lock_bh(&ecm_classifier_default_lock);
-			goto accel_no;
-		}
-		DEBUG_INFO("%px: Protocol: %d is non-TCP, prevailing_state ESTABLISHED\n", cdii, cdii->protocol);
-		goto check_delay;
-	}
-
-	/*
-	 * Check the TCP connection state, when the ct is NULL.
-	 * ct valid case was already checked in the ecm_nss{sfe}_ported_ipv4{6}_process functions.
-	 * If we are not established then we deny acceleration.
-	 */
-	ct = nf_ct_get(skb, &ctinfo);
-	if (!ct) {
-		DEBUG_TRACE("%px: No Conntrack found for packet, using ECM tracker state\n", cdii);
-
-		if (unlikely(prevailing_state == ECM_TRACKER_CONNECTION_STATE_ESTABLISHED)) {
-			DEBUG_INFO("%px: TCP prevailing_state ESTABLISHED\n", cdii);
-			goto check_delay;
-		}
-
-		DEBUG_INFO("%px: TCP prevailing_state: %d\n", cdii, prevailing_state);
-
-		spin_lock_bh(&ecm_classifier_default_lock);
-		if ((prevailing_state == ECM_TRACKER_CONNECTION_STATE_FAULT) || (prevailing_state == ECM_TRACKER_CONNECTION_STATE_CLOSED)) {
-			cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_TIMER_GROUP_NO_TOUCH;
+	/* CONSULTAR MARK CLASSIFIER (solo para SYN) */
+	ci = ecm_db_connection_serial_find_and_ref(cdii->ci_serial);
+	if (ci) {
+		mark_classi = ecm_db_connection_assigned_classifier_find_and_ref(ci, ECM_CLASSIFIER_TYPE_MARK);
+		if (mark_classi) {
+			mark_present = true;
+			// printk(KERN_INFO "DEFAULT: MARK classifier FOUND on connection\n");
+			mark_classi->last_process_response_get(mark_classi, &mark_pr);
+			mark_classi->deref(mark_classi);
+			// printk(KERN_INFO "DEFAULT: MARK accel_mode=%d, relevance=%d\n", mark_pr.accel_mode, mark_pr.relevance);
+
+			/* Si MARK dice NO_ACCEL (1), denegar */
+			if (mark_pr.relevance == ECM_CLASSIFIER_RELEVANCE_YES &&
+				mark_pr.accel_mode == ECM_CLASSIFIER_ACCELERATION_MODE_NO) {
+				mark_denies = true;
+				// printk(KERN_INFO "MARK: DEFAULT forcing slow path (mark denied)\n");
+				} else if (mark_pr.relevance == ECM_CLASSIFIER_RELEVANCE_YES &&
+					mark_pr.accel_mode == ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL) {
+					// printk(KERN_INFO "MARK: DEFAULT sees MARK says ACCEL\n");
+					} else {
+						// printk(KERN_INFO "MARK: DEFAULT sees MARK says DONT_CARE or MAYBE\n");
+					}
+		} else {
+			// printk(KERN_INFO "DEFAULT: MARK classifier NOT FOUND on connection\n");
 		}
-		goto accel_no;
+		ecm_db_connection_deref(ci);
+	} else {
+		// printk(KERN_INFO "DEFAULT: Connection NOT found for serial %u\n", cdii->ci_serial);
 	}
 
-	/*
-	* If the connection is shutting down do not manage it.
-	* state can not be SYN_SENT, SYN_RECV because connection is assured
-	* Not managed states: FIN_WAIT, CLOSE_WAIT, LAST_ACK, TIME_WAIT, CLOSE.
-	*/
-	spin_lock_bh(&ct->lock);
-	if (ct->proto.tcp.state != TCP_CONNTRACK_ESTABLISHED) {
-		spin_unlock_bh(&ct->lock);
-		DEBUG_TRACE("%px: Connection in termination state %#X\n", ct, ct->proto.tcp.state);
-		spin_lock_bh(&ecm_classifier_default_lock);
-		goto accel_no;
-	}
-	spin_unlock_bh(&ct->lock);
+	/* DECISION */
+	spin_lock_bh(&ecm_classifier_default_lock);
 
-check_delay:
-	/*
-	 * Should we delay the acceleration?
-	 */
-	if (ecm_classifier_accel_delay_pkts) {
-		if (!ecm_classifier_default_ready_for_accel(cdii, sender)) {
-			DEBUG_INFO("%px: connection is not ready for accel\n", cdii);
-			spin_lock_bh(&ecm_classifier_default_lock);
-			goto accel_no;
+	if (mark_present) {
+		if (mark_denies) {
+			/* MARK dice NO_ACCEL */
+			cdii->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_NO;
+			cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+
+			if (!cdii->decelerated) {
+				feci = ecm_db_connection_front_end_get_and_ref(ci);
+				if (feci) {
+					if (feci->decelerate) {
+						feci->decelerate(feci);
+						// printk(KERN_INFO "MARK: DECELERATED connection (destroyed NSS rule)\n");
+					}
+					ecm_front_end_connection_deref(feci);
+				}
+				cdii->decelerated = true;
+			}
+			// printk(KERN_INFO "DEFAULT: MARK says NO_ACCEL, decision = NO_ACCEL\n");
+		} else {
+			/* MARK dice ACCEL o DONT_CARE → DEFAULT acelera */
+			cdii->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL;
+			cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+			// printk(KERN_INFO "DEFAULT: MARK says ACCEL or DONT_CARE, decision = ACCEL\n");
 		}
-		DEBUG_INFO("%px: connection is ready for accel\n", cdii);
-	}
-
-	/*
-	 * Return the process response
-	 */
-	spin_lock_bh(&ecm_classifier_default_lock);
-	cdii->process_response.accel_mode = ecm_classifier_default_accel_mode;
-	cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+	} else {
+		/* MARK NO está presente, DEFAULT decide por sí mismo */
+		cdii->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL;
+		cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+		// printk(KERN_INFO "DEFAULT: MARK NOT present, decision = ACCEL\n");
+	}
+	/* DEBUG: Marcar como clasificada para no volver a procesar */
+	cdii->classified = true;
+	/* Establecer YES para que NSS sepa que debe acelerar */
+	cdii->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
 	*process_response = cdii->process_response;
-	spin_unlock_bh(&ecm_classifier_default_lock);
-	return;
 
-accel_no:
-	cdii->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_NO;
-	cdii->process_response.process_actions |= ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
-	*process_response = cdii->process_response;
 	spin_unlock_bh(&ecm_classifier_default_lock);
+
+	// printk(KERN_INFO "DEFAULT: Decision = %s, classified=%d, mark_present=%d\n", cdii->process_response.accel_mode == ECM_CLASSIFIER_ACCELERATION_MODE_NO ? "NO_ACCEL" : "ACCEL", cdii->classified, mark_present);
 }
 
 /*
@@ -481,7 +339,7 @@
  *	Get result code returned by the last process call
  */
 static void ecm_classifier_default_last_process_response_get(struct ecm_classifier_instance *aci,
-							struct ecm_classifier_process_response *process_response)
+															 struct ecm_classifier_process_response *process_response)
 {
 	struct ecm_classifier_default_internal_instance *cdii;
 	cdii = (struct ecm_classifier_default_internal_instance *)aci;
@@ -598,9 +456,6 @@
 		return result;
 	}
 
-	/*
-	 * Output our last process response
-	 */
 	if ((result = ecm_classifier_process_response_state_get(sfi, &process_response))) {
 		return result;
 	}
@@ -609,9 +464,6 @@
 		return result;
 	}
 
-	/*
-	 * Output our tracker state
-	 */
 	if ((result = cdii->ti->state_text_get(cdii->ti, sfi))) {
 		return result;
 	}
@@ -626,25 +478,19 @@
 
 /*
  * ecm_classifier_default_instance_alloc()
- *	Allocate an instance of the default classifier
+ * Allocate an instance of the default classifier
  */
 struct ecm_classifier_default_instance *ecm_classifier_default_instance_alloc(struct ecm_db_connection_instance *ci, int protocol, ecm_db_direction_t dir, int from_port, int to_port)
 {
 	struct ecm_classifier_default_internal_instance *cdii;
 	struct ecm_classifier_default_instance *cdi;
 
-	/*
-	 * Allocate the instance
-	 */
 	cdii = (struct ecm_classifier_default_internal_instance *)kzalloc(sizeof(struct ecm_classifier_default_internal_instance), GFP_ATOMIC | __GFP_NOWARN);
 	if (!cdii) {
 		DEBUG_WARN("Failed to allocate default instance\n");
 		return NULL;
 	}
 
-	/*
-	 * Allocate a tracker for state etc.
-	 */
 	if (protocol == IPPROTO_TCP) {
 		DEBUG_TRACE("%px: Alloc tracker for TCP connection: %px\n", cdii, ci);
 		cdii->ti = (struct ecm_tracker_instance *)ecm_tracker_tcp_alloc();
@@ -679,14 +525,10 @@
 	cdii->ci_serial = ecm_db_connection_serial_get(ci);
 	cdii->protocol = protocol;
 
-	/*
-	 * We are always relevant to the connection
-	 */
 	cdii->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
+	cdii->classified = false;
+	cdii->decelerated = false;
 
-	/*
-	 * Using the connection direction identify egress and ingress host addresses
-	 */
 	if (dir == ECM_DB_DIRECTION_INGRESS_NAT) {
 		cdii->ingress_sender = ECM_TRACKER_SENDER_TYPE_SRC;
 		cdii->egress_sender = ECM_TRACKER_SENDER_TYPE_DEST;
@@ -696,15 +538,9 @@
 	}
 	DEBUG_TRACE("%px: Ingress sender = %d egress sender = %d\n", cdii, cdii->ingress_sender, cdii->egress_sender);
 
-	/*
-	 * Methods specific to the default classifier
-	 */
 	cdi = (struct ecm_classifier_default_instance *)cdii;
 	cdi->tracker_get_and_ref = ecm_classifier_tracker_get_and_ref;
 
-	/*
-	 * Methods generic to all classifiers.
-	 */
 	cdi->base.process = ecm_classifier_default_process;
 	cdi->base.sync_from_v4 = ecm_classifier_default_sync_from_v4;
 	cdi->base.sync_to_v4 = ecm_classifier_default_sync_to_v4;
@@ -714,17 +550,14 @@
 	cdi->base.reclassify_allowed = ecm_classifier_default_reclassify_allowed;
 	cdi->base.reclassify = ecm_classifier_default_reclassify;
 	cdi->base.last_process_response_get = ecm_classifier_default_last_process_response_get;
-#ifdef ECM_STATE_OUTPUT_ENABLE
+	#ifdef ECM_STATE_OUTPUT_ENABLE
 	cdi->base.state_get = ecm_classifier_default_state_get;
-#endif
+	#endif
 	cdi->base.ref = ecm_classifier_default_ref;
 	cdi->base.deref = ecm_classifier_default_deref;
 
 	spin_lock_bh(&ecm_classifier_default_lock);
 
-	/*
-	 * Final check if we are pending termination
-	 */
 	if (ecm_classifier_default_terminate_pending) {
 		spin_unlock_bh(&ecm_classifier_default_lock);
 		DEBUG_INFO("%px: Terminating\n", ci);
@@ -733,9 +566,6 @@
 		return NULL;
 	}
 
-	/*
-	 * Increment stats
-	 */
 	ecm_classifier_default_count++;
 	DEBUG_ASSERT(ecm_classifier_default_count > 0, "%px: ecm_classifier_default_count wrap\n", cdii);
 	spin_unlock_bh(&ecm_classifier_default_lock);
@@ -761,27 +591,27 @@
 	}
 
 	if (!ecm_debugfs_create_u32("enabled", S_IRUGO | S_IWUSR, ecm_classifier_default_dentry,
-					(u32 *)&ecm_classifier_default_enabled)) {
-		DEBUG_ERROR("Failed to create ecm deafult classifier enabled file in debugfs\n");
-		debugfs_remove_recursive(ecm_classifier_default_dentry);
-		return -1;
-	}
+		(u32 *)&ecm_classifier_default_enabled)) {
+		DEBUG_ERROR("Failed to create ecm default classifier enabled file in debugfs\n");
+	debugfs_remove_recursive(ecm_classifier_default_dentry);
+	return -1;
+		}
 
-	if (!ecm_debugfs_create_u32("accel_mode", S_IRUGO | S_IWUSR, ecm_classifier_default_dentry,
-					(u32 *)&ecm_classifier_default_accel_mode)) {
-		DEBUG_ERROR("Failed to create ecm deafult classifier accel_mode file in debugfs\n");
+		if (!ecm_debugfs_create_u32("accel_mode", S_IRUGO | S_IWUSR, ecm_classifier_default_dentry,
+			(u32 *)&ecm_classifier_default_accel_mode)) {
+			DEBUG_ERROR("Failed to create ecm default classifier accel_mode file in debugfs\n");
 		debugfs_remove_recursive(ecm_classifier_default_dentry);
 		return -1;
-	}
+			}
 
-	if (!ecm_debugfs_create_u32("accel_delay_pkts", S_IRUGO | S_IWUSR, ecm_classifier_default_dentry,
-					(u32 *)&ecm_classifier_accel_delay_pkts)) {
-		DEBUG_ERROR("Failed to create accel delay packet counts in debugfs\n");
-		debugfs_remove_recursive(ecm_classifier_default_dentry);
-		return -1;
-	}
+			if (!ecm_debugfs_create_u32("accel_delay_pkts", S_IRUGO | S_IWUSR, ecm_classifier_default_dentry,
+				(u32 *)&ecm_classifier_accel_delay_pkts)) {
+				DEBUG_ERROR("Failed to create accel delay packet counts in debugfs\n");
+			debugfs_remove_recursive(ecm_classifier_default_dentry);
+			return -1;
+				}
 
-	return 0;
+				return 0;
 }
 EXPORT_SYMBOL(ecm_classifier_default_init);
 
@@ -795,9 +625,6 @@
 	ecm_classifier_default_terminate_pending = true;
 	spin_unlock_bh(&ecm_classifier_default_lock);
 
-	/*
-	 * Remove the debugfs files recursively.
-	 */
 	if (ecm_classifier_default_dentry) {
 		debugfs_remove_recursive(ecm_classifier_default_dentry);
 	}
diff -Naur "a/ecm_classifier_mark.c" b/ecm_classifier_mark.c
--- "a/ecm_classifier_mark.c"	2024-11-06 13:33:31.000000000 +0100
+++ b/ecm_classifier_mark.c	2026-08-05 16:57:14.811991793 +0200
@@ -1,5 +1,4 @@
 /*
- **************************************************************************
  * Copyright (c) 2018, 2020-2021, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -11,16 +10,16 @@
  * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
  * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
- **************************************************************************
  */
 
 /*
  * Mark Classifier.
- * This module provides an interface to customer's external module to interract with the ECM.
+ * This module provides an interface to customer's external module to interact with the ECM.
  * It allows the customer to make smart decisions on the packets before offloading the connection to the
  * acceleration engine.
  */
 
+#include <linux/printk.h>
 #include <linux/version.h>
 #include <linux/types.h>
 #include <linux/ip.h>
@@ -33,7 +32,6 @@
 #include <linux/inet.h>
 #include <linux/in.h>
 #include <linux/etherdevice.h>
-
 /*
  * Debug output levels
  * 0 = OFF
@@ -62,68 +60,74 @@
 #include "ecm_front_end_ipv6.h"
 #endif
 
-/*
- * Magic numbers
- */
 #define ECM_CLASSIFIER_MARK_INSTANCE_MAGIC 0x2567
 
 /*
  * struct ecm_classifier_mark_instance
- * 	State per connection for MARK classifier
+ *	State per connection for MARK classifier
  */
 struct ecm_classifier_mark_instance {
-	struct ecm_classifier_instance base;			/* Base type */
-
-	uint32_t ci_serial;					/* RO: Serial of the connection */
-
-	struct ecm_classifier_mark_instance *next;		/* Next classifier state instance (for accouting and reporting purposes) */
-	struct ecm_classifier_mark_instance *prev;		/* Next classifier state instance (for accouting and reporting purposes) */
-
-	uint32_t mark;						/* Mark value which this classifier extracted from the buffer */
+	struct ecm_classifier_instance base;
+	uint32_t ci_serial;
+	struct ecm_classifier_mark_instance *next;
+	struct ecm_classifier_mark_instance *prev;
+	uint32_t mark;
 	struct ecm_classifier_process_response process_response;
-								/* Last process response computed */
-	int refs;						/* Integer to trap we never go negative */
-#if (DEBUG_LEVEL > 0)
+	int refs;
+	bool decelerated;
+	#if (DEBUG_LEVEL > 0)
 	uint16_t magic;
-#endif
+	#endif
 };
 
-/*
- * Operational control. By default it is disabled.
- */
-static int ecm_classifier_mark_enabled;			/* Operational behaviour */
-
-/*
- * Management thread control
- */
-static bool ecm_classifier_mark_terminate_pending;	/* True when the user wants us to terminate */
+static int ecm_classifier_mark_enabled = 1;
+static uint32_t ecm_classifier_mark_deny_value = 0x100;
+static uint32_t ecm_classifier_mark_mode = 0;
+static bool ecm_classifier_mark_terminate_pending;
 
-/*
- * Debugfs dentry object.
- */
 static struct dentry *ecm_classifier_mark_dentry;
-
-/*
- * Locking of the classifier structures
- */
-static DEFINE_SPINLOCK(ecm_classifier_mark_lock);			/* Protect SMP access. */
-
-/*
- * List of our classifier instances
- */
+static DEFINE_SPINLOCK(ecm_classifier_mark_lock);
 static struct ecm_classifier_mark_instance *ecm_classifier_mark_instances = NULL;
-								/* list of all active instances */
-static int ecm_classifier_mark_count = 0;			/* Tracks number of instances allocated */
+static int ecm_classifier_mark_count = 0;
 
-/*
- * Callbacks to the external modules.
- */
 ecm_classifier_mark_get_callback_t mark_get_cb[ECM_CLASSIFIER_MARK_TYPE_MAX] = { NULL };
-								/* Callbacks which gets the mark from the packets */
 ecm_classifier_mark_sync_to_ipv4_callback_t sync_to_ipv4_cb[ECM_CLASSIFIER_MARK_TYPE_MAX] = { NULL };
-								/* Callbacks which syncs the IPv4 stats to the registered external subsystems */
 ecm_classifier_mark_sync_to_ipv6_callback_t sync_to_ipv6_cb[ECM_CLASSIFIER_MARK_TYPE_MAX] = { NULL };
-								/* Callbacks which syncs the IPv6 stats to the registered external subsystems */
+
+/*
+ * Default mark callbacks
+ */
+static ecm_classifier_mark_result_t ecm_classifier_mark_default_get(struct sk_buff *skb, uint32_t *mark)
+{
+	*mark = skb->mark;
+	// printk(KERN_INFO "MARK: default_get called - skb->mark = 0x%x\n", *mark);
+
+	/* Si la marca es 0x100, DENEGAR aceleración explícitamente */
+	if (*mark == 0x100) {
+		// printk(KERN_INFO "MARK: Mark 0x100 detected - DENY acceleration\n");
+		return ECM_CLASSIFIER_MARK_RESULT_NOT_RELEVANT;
+	}
+
+	// printk(KERN_INFO "MARK: Allowing accel for mark 0x%x\n", *mark);
+	return ECM_CLASSIFIER_MARK_RESULT_SUCCESS;
+}
+
+static void ecm_classifier_mark_default_sync_ipv4(uint32_t mark,
+												  __be32 src_ip, int src_port,
+												  __be32 dest_ip, int dest_port,
+												  int protocol)
+{
+	DEBUG_TRACE("MARK Sync IPv4: mark=0x%x %pI4:%d -> %pI4:%d proto=%d\n",
+				mark, &src_ip, src_port, &dest_ip, dest_port, protocol);
+}
+
+static void ecm_classifier_mark_default_sync_ipv6(uint32_t mark,
+												  struct in6_addr *src_ip, int src_port,
+												  struct in6_addr *dest_ip, int dest_port,
+												  int protocol)
+{
+	DEBUG_TRACE("MARK Sync IPv6: mark=0x%x proto=%d\n", mark, protocol);
+}
 
 /*
  * _ecm_classifier_mark_ref()
@@ -171,15 +175,9 @@
 		return refs;
 	}
 
-	/*
-	 * Object to be destroyed
-	 */
 	ecm_classifier_mark_count--;
 	DEBUG_ASSERT(ecm_classifier_mark_count >= 0, "%px: ecm_classifier_mark_count wrap\n", ecmi);
 
-	/*
-	 * UnLink the instance from our list
-	 */
 	if (ecmi->next) {
 		ecmi->next->prev = ecmi->prev;
 	}
@@ -193,9 +191,6 @@
 	ecmi->prev = NULL;
 	spin_unlock_bh(&ecm_classifier_mark_lock);
 
-	/*
-	 * Final
-	 */
 	DEBUG_INFO("%px: Final mark classifier instance\n", ecmi);
 	kfree(ecmi);
 
@@ -205,12 +200,10 @@
 /*
  * ecm_classifier_mark_process()
  *	Process new packet
- *
- * NOTE: This function would only ever be called if all other classifiers have failed.
  */
 static void ecm_classifier_mark_process(struct ecm_classifier_instance *aci, ecm_tracker_sender_type_t sender,
-									struct ecm_tracker_ip_header *ip_hdr, struct sk_buff *skb,
-									struct ecm_classifier_process_response *process_response)
+										struct ecm_tracker_ip_header *ip_hdr, struct sk_buff *skb,
+										struct ecm_classifier_process_response *process_response)
 {
 	struct ecm_classifier_mark_instance *ecmi = (struct ecm_classifier_mark_instance *)aci;
 	ecm_classifier_mark_result_t result;
@@ -222,157 +215,172 @@
 	uint32_t mark = 0;
 	ecm_classifier_mark_get_callback_t cb = NULL;
 	ecm_classifier_mark_type_t type = ECM_CLASSIFIER_MARK_TYPE_L2_ENCAP;
-			/* TODO: By default we use this type. In the future the type will be gotten from the connection instance*/
+
+	// printk(KERN_INFO "MARK: process() ENTERED for serial %u, decelerated=%d\n", ecmi->ci_serial, ecmi->decelerated);
+
+	/* ===== Si ya decelerated, NO hacer nada ===== */
+	if (ecmi->decelerated) {
+		// printk(KERN_INFO "MARK: Already decelerated, ignoring\n");
+		*process_response = ecmi->process_response;
+		return;
+	}
 
 	DEBUG_CHECK_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC, "%px: invalid state magic\n", ecmi);
 
 	/*
-	 * Not relevant to the connection if not enabled.
+	 * CASO 1: MARK no está habilitado → DESASIGNAR
 	 */
 	if (unlikely(!ecm_classifier_mark_enabled)) {
-		/*
-		 * Not relevant.
-		 */
 		DEBUG_WARN("%px: Mark classifier is not enabled.\n", aci);
-		goto not_relevant;
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecmi->decelerated = true;
+		return;
 	}
 
 	/*
-	 * Get connection
+	 * Obtener la conexión
 	 */
 	ci = ecm_db_connection_serial_find_and_ref(ecmi->ci_serial);
 	if (!ci) {
-		/*
-		 * Connection has gone from under us
-		 */
 		DEBUG_WARN("%px: connection instance gone while processing classifier.\n", aci);
-		goto not_relevant;
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecmi->decelerated = true;
+		return;
 	}
 
 	/*
-	 * Get the L2 encap protocol of the connection.
+	 * CASO 2: mode == 2 (allow all) → PERMITIR ACCEL, DESASIGNAR
 	 */
-	l2_encap_proto = ecm_db_connection_l2_encap_proto_get(ci);
-	if (!l2_encap_proto) {
-		/*
-		 * The caller is not interested in the inspection. Not relevant.
-		 */
-		DEBUG_WARN("%px: L2 encap protocol is zero.\n", aci);
+	if (ecm_classifier_mark_mode == 2) {
+		DEBUG_TRACE("%px: MARK mode=2 (allow all)\n", aci);
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
+		ecmi->process_response.process_actions = ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+		ecmi->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL;
+		*process_response = ecmi->process_response;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecmi->decelerated = true;
 		ecm_db_connection_deref(ci);
-		goto not_relevant;
+		return;
 	}
 
 	/*
-	 * Is there an external callback to get the mark value from the packet?
+	 * CASO 3: mode == 1 (deny all) → DENEGAR ACCEL, DESASIGNAR
 	 */
-	rcu_read_lock();
-	cb = rcu_dereference(mark_get_cb[type]);
-	if (!cb) {
-		/*
-		 * Not relevant.
-		 */
-		rcu_read_unlock();
-		DEBUG_WARN("%px: No inspection callback set with type: %d.\n", aci, type);
+	if (ecm_classifier_mark_mode == 1) {
+		DEBUG_TRACE("%px: MARK mode=1 (deny all)\n", aci);
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
+		ecmi->process_response.process_actions = ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+		ecmi->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_NO;
+		*process_response = ecmi->process_response;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecmi->decelerated = true;
 		ecm_db_connection_deref(ci);
-		goto not_relevant;
+		return;
 	}
 
 	/*
-	 * Before giving the skb to the external callback for getting the mark value,
-	 * let's adjust the L2 encap header.
-	 * If the skb's protocol is not equal to the connection's L2 encap protocol,
-	 * this means the skb's L2 header is pulled off. So, let's push it back.
+	 * Obtener el protocolo L2 encap
 	 */
-	if (l2_encap_proto != ntohs(skb->protocol)) {
-		DEBUG_TRACE("%px: Connections L2 encap protocol: %x is different than skb's protocol: %x, need header adjustment.\n",
-			    aci, l2_encap_proto, ntohs(skb->protocol));
-		l2_encap_len = ecm_front_end_l2_encap_header_len(l2_encap_proto);
-		skb_header_adjusted = true;
-		ecm_front_end_push_l2_encap_header(skb, l2_encap_len);
-		skb_proto = skb->protocol;
-		skb->protocol = htons(l2_encap_proto);
-	}
+	// printk(KERN_INFO "MARK: l2_encap_proto check - ci=%px\n", ci);
+	// l2_encap_proto = ecm_db_connection_l2_encap_proto_get(ci);
+	// printk(KERN_INFO "MARK: l2_encap_proto = %d\n", l2_encap_proto);
+	// if (!l2_encap_proto) {
+	// 	DEBUG_WARN("%px: L2 encap protocol is zero.\n", aci);
+	// 	ecm_db_connection_deref(ci);
+	// 	spin_lock_bh(&ecm_classifier_mark_lock);
+	// 	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
+	// 	spin_unlock_bh(&ecm_classifier_mark_lock);
+	// 	ecmi->decelerated = true;
+	// 	return;
+	// }
 
 	/*
-	 * Call the external callback and get the result.
+	 * Obtener la marca del paquete
 	 */
-	result = cb(skb, &mark);
-	rcu_read_unlock();
+	rcu_read_lock();
+	cb = rcu_dereference(mark_get_cb[type]);
+	if (!cb) {
+		rcu_read_unlock();
+		result = ecm_classifier_mark_default_get(skb, &mark);
+	} else {
+		if (l2_encap_proto != ntohs(skb->protocol)) {
+			l2_encap_len = ecm_front_end_l2_encap_header_len(l2_encap_proto);
+			skb_header_adjusted = true;
+			ecm_front_end_push_l2_encap_header(skb, l2_encap_len);
+			skb_proto = skb->protocol;
+			skb->protocol = htons(l2_encap_proto);
+		}
 
-	/*
-	 * Pull the skb's L2 header, if it is adjusted in this classifier.
-	 */
-	if (skb_header_adjusted) {
-		DEBUG_TRACE("%px: Adjust the L2 header back.\n", aci);
-		ecm_front_end_pull_l2_encap_header(skb, l2_encap_len);
-		skb->protocol = skb_proto;
+		result = cb(skb, &mark);
+		rcu_read_unlock();
+
+		if (skb_header_adjusted) {
+			ecm_front_end_pull_l2_encap_header(skb, l2_encap_len);
+			skb->protocol = skb_proto;
+		}
 	}
 
 	/*
-	 * Handle the result
+	 * ================================================================
+	 * DECISIÓN BASADA EN LA MARCA
+	 * ================================================================
 	 */
-	switch (result) {
-	case ECM_CLASSIFIER_MARK_RESULT_NOT_YET:
-		/*
-		 * Deny accel but don't change the permit state - we try again later
-		 */
-		DEBUG_WARN("%px: External callback has not decided yet.\n", aci);
-		goto deny_accel;
-	case ECM_CLASSIFIER_MARK_RESULT_NOT_RELEVANT:
-		/*
-		 * The callback told us that the classifier is not relevant to this connection.
-		 */
-		DEBUG_WARN("%px: External callback decided not relevant.\n", aci);
-		ecm_db_connection_deref(ci);
-		goto not_relevant;
-	case ECM_CLASSIFIER_MARK_RESULT_SUCCESS:
-
-		/*
-		 * Copy the returned mark value to the connection and the classifier instance.
-		 * We copy the mark value to the classifier instance as well here for quick access
-		 * in the sync functions.
-		 */
-		DEBUG_WARN("%px: External callback returned the mark value: %x\n", aci, mark);
+	if (result == ECM_CLASSIFIER_MARK_RESULT_SUCCESS) {
+		/* Marca OK → permitir ACCEL, DESASIGNAR */
+		// printk(KERN_INFO "MARK: SUCCESS for serial %u, mark=0x%x -> ACCEL\n", ecmi->ci_serial, mark);
 		ecm_db_connection_mark_set(ci, mark);
 		ecmi->mark = mark;
-		break;
-	default:
-		DEBUG_ASSERT(false, "Unhandled result: %d\n", result);
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
+		ecmi->process_response.process_actions = ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+		ecmi->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL;
+		*process_response = ecmi->process_response;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecmi->decelerated = true;
+		ecm_db_connection_deref(ci);
+		return;
 	}
 
-	/*
-	 * Acceleration is permitted
-	 */
-	spin_lock_bh(&ecm_classifier_mark_lock);
-	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
-	ecmi->process_response.process_actions = ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
-	ecmi->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_ACCEL;
-	*process_response = ecmi->process_response;
-	spin_unlock_bh(&ecm_classifier_mark_lock);
-	ecm_db_connection_deref(ci);
-
-	return;
+	if (result == ECM_CLASSIFIER_MARK_RESULT_NOT_RELEVANT) {
+		/* DEBUG De NOT_RELEVANT a DENEGAR ACCEL */
+		/* La marca 0x100 fue detectada, forzar NO_ACCEL */
+		// printk(KERN_INFO "MARK: NOT_RELEVANT for serial %u, mark=0x%x -> DENY ACCEL (mark denied)\n", ecmi->ci_serial, mark);
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
+		ecmi->process_response.process_actions = ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
+		ecmi->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_NO;
+		*process_response = ecmi->process_response;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecmi->decelerated = true;
+		ecm_db_connection_deref(ci);
+		return;
+	}
 
-not_relevant:
-	/*
-	 * ecm_classifier_mark_lock MUST be held
-	 */
-	spin_lock_bh(&ecm_classifier_mark_lock);
-	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_NO;
-	*process_response = ecmi->process_response;
-	spin_unlock_bh(&ecm_classifier_mark_lock);
-	return;
+	/* NOT_YET: esperar */
+	if (result == ECM_CLASSIFIER_MARK_RESULT_NOT_YET) {
+		// (KERN_INFO "MARK: NOT_YET for serial %u, mark=0x%x -> waiting\n", ecmi->ci_serial, mark);
+		spin_lock_bh(&ecm_classifier_mark_lock);
+		ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_MAYBE;
+		spin_unlock_bh(&ecm_classifier_mark_lock);
+		ecm_db_connection_deref(ci);
+		return;
+	}
 
-deny_accel:
-	/*
-	 * ecm_classifier_mark_lock MUST be held
-	 */
+	/* Fallback: si algo no esperado ocurre, denegar por defecto */
+	// printk(KERN_INFO "MARK: UNKNOWN result for serial %u, mark=0x%x -> DENY ACCEL\n", ecmi->ci_serial, mark);
 	spin_lock_bh(&ecm_classifier_mark_lock);
 	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
 	ecmi->process_response.process_actions = ECM_CLASSIFIER_PROCESS_ACTION_ACCEL_MODE;
 	ecmi->process_response.accel_mode = ECM_CLASSIFIER_ACCELERATION_MODE_NO;
 	*process_response = ecmi->process_response;
 	spin_unlock_bh(&ecm_classifier_mark_lock);
+	ecmi->decelerated = true;
 	ecm_db_connection_deref(ci);
 }
 
@@ -384,7 +392,6 @@
 {
 	struct ecm_classifier_mark_instance *ecmi;
 	ecmi = (struct ecm_classifier_mark_instance *)aci;
-
 	DEBUG_CHECK_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC, "%px: magic failed", ecmi);
 	return ECM_CLASSIFIER_TYPE_MARK;
 }
@@ -397,7 +404,6 @@
 {
 	struct ecm_classifier_mark_instance *ecmi;
 	ecmi = (struct ecm_classifier_mark_instance *)aci;
-
 	DEBUG_CHECK_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC, "%px: magic failed", ecmi);
 	return true;
 }
@@ -412,12 +418,11 @@
 	ecmi = (struct ecm_classifier_mark_instance *)aci;
 	DEBUG_CHECK_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC, "%px: magic failed", ecmi);
 
-	/*
-	 * Revert back to MAYBE relevant - we will evaluate when we get the next process() call.
-	 */
 	spin_lock_bh(&ecm_classifier_mark_lock);
-	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_MAYBE;
+	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
 	spin_unlock_bh(&ecm_classifier_mark_lock);
+
+	// printk(KERN_INFO "MARK: reclassify() called, relevance set to YES\n");
 }
 
 /*
@@ -425,7 +430,7 @@
  *	Get result code returned by the last process call
  */
 static void ecm_classifier_mark_last_process_response_get(struct ecm_classifier_instance *aci,
-							struct ecm_classifier_process_response *process_response)
+														  struct ecm_classifier_process_response *process_response)
 {
 	struct ecm_classifier_mark_instance *ecmi;
 	ecmi = (struct ecm_classifier_mark_instance *)aci;
@@ -452,21 +457,14 @@
 	struct ecm_db_connection_instance *ci;
 	ecm_classifier_mark_sync_to_ipv4_callback_t cb = NULL;
 	ecm_classifier_mark_type_t type = ECM_CLASSIFIER_MARK_TYPE_L2_ENCAP;
-			/* TODO: By default we use this type. In the future the type will be gotten from the connection instance */
 	struct ecm_classifier_mark_instance *ecmi = (struct ecm_classifier_mark_instance *)aci;
+
 	DEBUG_CHECK_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC, "%px: magic failed", ecmi);
 
-	/*
-	 * Nothing to update.
-	 * We only care about flows that are actively being accelerated.
-	 */
 	if (!(sync->tx_packet_count[ECM_CONN_DIR_FLOW] || sync->tx_packet_count[ECM_CONN_DIR_RETURN])) {
 		return;
 	}
 
-	/*
-	 * Handle only the stats sync. We don't care about the evict or flush syncs.
-	 */
 	if (sync->reason != ECM_FRONT_END_IPV4_RULE_SYNC_REASON_STATS) {
 		return;
 	}
@@ -528,21 +526,14 @@
 	struct ecm_db_connection_instance *ci;
 	ecm_classifier_mark_sync_to_ipv6_callback_t cb = NULL;
 	ecm_classifier_mark_type_t type = ECM_CLASSIFIER_MARK_TYPE_L2_ENCAP;
-			/* TODO: By default we use this type. In the future the type will be gotten from the connection instance */
 	struct ecm_classifier_mark_instance *ecmi = (struct ecm_classifier_mark_instance *)aci;
+
 	DEBUG_CHECK_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC, "%px: magic failed", ecmi);
 
-	/*
-	 * Nothing to update.
-	 * We only care about flows that are actively being accelerated.
-	 */
 	if (!(sync->tx_packet_count[ECM_CONN_DIR_FLOW] || sync->tx_packet_count[ECM_CONN_DIR_RETURN])) {
 		return;
 	}
 
-	/*
-	 * Handle only the stats sync. We don't care about the evict or flush syncs.
-	 */
 	if (sync->reason != ECM_FRONT_END_IPV4_RULE_SYNC_REASON_STATS) {
 		return;
 	}
@@ -590,7 +581,7 @@
 #ifdef ECM_STATE_OUTPUT_ENABLE
 /*
  * ecm_classifier_mark_state_get()
- *	Gets the state of this classfier and outputs it to debugfs.
+ *	Gets the state of this classifier and outputs it to debugfs.
  */
 static int ecm_classifier_mark_state_get(struct ecm_classifier_instance *ci, struct ecm_state_file_instance *sfi)
 {
@@ -615,9 +606,6 @@
 		return result;
 	}
 
-	/*
-	 * Output our last process response
-	 */
 	if ((result = ecm_classifier_process_response_state_get(sfi, &process_response))) {
 		return result;
 	}
@@ -634,9 +622,6 @@
 {
 	struct ecm_classifier_mark_instance *ecmi;
 
-	/*
-	 * Allocate the instance
-	 */
 	ecmi = (struct ecm_classifier_mark_instance *)kzalloc(sizeof(struct ecm_classifier_mark_instance), GFP_ATOMIC | __GFP_NOWARN);
 	if (!ecmi) {
 		DEBUG_WARN("i%px: Failed to allocate Mark Classifier instance\n", ci);
@@ -646,9 +631,6 @@
 	DEBUG_SET_MAGIC(ecmi, ECM_CLASSIFIER_MARK_INSTANCE_MAGIC);
 	ecmi->refs = 1;
 
-	/*
-	 * Methods generic to all classifiers.
-	 */
 	ecmi->base.process = ecm_classifier_mark_process;
 	ecmi->base.sync_from_v4 = ecm_classifier_mark_sync_from_v4;
 	ecmi->base.sync_to_v4 = ecm_classifier_mark_sync_to_v4;
@@ -658,19 +640,17 @@
 	ecmi->base.reclassify_allowed = ecm_classifier_mark_reclassify_allowed;
 	ecmi->base.reclassify = ecm_classifier_mark_reclassify;
 	ecmi->base.last_process_response_get = ecm_classifier_mark_last_process_response_get;
-#ifdef ECM_STATE_OUTPUT_ENABLE
+	#ifdef ECM_STATE_OUTPUT_ENABLE
 	ecmi->base.state_get = ecm_classifier_mark_state_get;
-#endif
+	#endif
 	ecmi->base.ref = ecm_classifier_mark_ref;
 	ecmi->base.deref = ecm_classifier_mark_deref;
 	ecmi->ci_serial = ecm_db_connection_serial_get(ci);
 
 	ecmi->process_response.process_actions = 0;
-	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_MAYBE;
+	ecmi->process_response.relevance = ECM_CLASSIFIER_RELEVANCE_YES;
+	ecmi->decelerated = false;
 
-	/*
-	 * Final check if we are pending termination
-	 */
 	spin_lock_bh(&ecm_classifier_mark_lock);
 	if (ecm_classifier_mark_terminate_pending) {
 		spin_unlock_bh(&ecm_classifier_mark_lock);
@@ -679,18 +659,12 @@
 		return NULL;
 	}
 
-	/*
-	 * Link the new instance into our list at the head
-	 */
 	ecmi->next = ecm_classifier_mark_instances;
 	if (ecm_classifier_mark_instances) {
 		ecm_classifier_mark_instances->prev = ecmi;
 	}
 	ecm_classifier_mark_instances = ecmi;
 
-	/*
-	 * Increment stats
-	 */
 	ecm_classifier_mark_count++;
 	DEBUG_ASSERT(ecm_classifier_mark_count > 0, "%px: ecm_classifier_mark_count wrap for instance: %px\n", ci, ecmi);
 	spin_unlock_bh(&ecm_classifier_mark_lock);
@@ -705,9 +679,9 @@
  *	Register external callbacks.
  */
 int ecm_classifier_mark_register_callbacks(ecm_classifier_mark_type_t type,
-					   ecm_classifier_mark_get_callback_t mark_get,
-					   ecm_classifier_mark_sync_to_ipv4_callback_t sync_to_ipv4,
-					   ecm_classifier_mark_sync_to_ipv6_callback_t sync_to_ipv6)
+										   ecm_classifier_mark_get_callback_t mark_get,
+										   ecm_classifier_mark_sync_to_ipv4_callback_t sync_to_ipv4,
+										   ecm_classifier_mark_sync_to_ipv6_callback_t sync_to_ipv6)
 {
 	spin_lock_bh(&ecm_classifier_mark_lock);
 	if (ecm_classifier_mark_terminate_pending) {
@@ -718,7 +692,7 @@
 	spin_unlock_bh(&ecm_classifier_mark_lock);
 
 	DEBUG_ASSERT(!mark_get_cb[type] && !sync_to_ipv4_cb[type] && !sync_to_ipv6_cb[type],
-			"Mark callbacks are already registered\n");
+				 "Mark callbacks are already registered\n");
 	rcu_assign_pointer(mark_get_cb[type], mark_get);
 	rcu_assign_pointer(sync_to_ipv4_cb[type], sync_to_ipv4);
 	rcu_assign_pointer(sync_to_ipv6_cb[type], sync_to_ipv6);
@@ -736,7 +710,6 @@
 	rcu_assign_pointer(mark_get_cb[type], NULL);
 	rcu_assign_pointer(sync_to_ipv4_cb[type], NULL);
 	rcu_assign_pointer(sync_to_ipv6_cb[type], NULL);
-
 }
 EXPORT_SYMBOL(ecm_classifier_mark_unregister_callbacks);
 
@@ -745,6 +718,8 @@
  */
 int ecm_classifier_mark_init(struct dentry *dentry)
 {
+	// printk(KERN_INFO "MARK: INIT - Starting initialization\n");
+
 	DEBUG_INFO("Mark classifier Module init\n");
 
 	ecm_classifier_mark_dentry = debugfs_create_dir("ecm_classifier_mark", dentry);
@@ -754,13 +729,36 @@
 	}
 
 	if (!ecm_debugfs_create_u32("enabled", S_IRUGO | S_IWUSR, ecm_classifier_mark_dentry,
-					(u32 *)&ecm_classifier_mark_enabled)) {
+		(u32 *)&ecm_classifier_mark_enabled)) {
 		DEBUG_ERROR("Failed to create mark enabled file in debugfs\n");
+	debugfs_remove_recursive(ecm_classifier_mark_dentry);
+	return -1;
+		}
+
+		if (!ecm_debugfs_create_u32("deny_mark", S_IRUGO | S_IWUSR, ecm_classifier_mark_dentry,
+			(u32 *)&ecm_classifier_mark_deny_value)) {
+			DEBUG_ERROR("Failed to create deny_mark file in debugfs\n");
 		debugfs_remove_recursive(ecm_classifier_mark_dentry);
 		return -1;
-	}
+			}
 
-	return 0;
+			if (!ecm_debugfs_create_u32("mode", S_IRUGO | S_IWUSR, ecm_classifier_mark_dentry,
+				(u32 *)&ecm_classifier_mark_mode)) {
+					DEBUG_ERROR("Failed to create mode file in debugfs\n");
+					debugfs_remove_recursive(ecm_classifier_mark_dentry);
+					return -1;
+				}
+
+				rcu_assign_pointer(mark_get_cb[ECM_CLASSIFIER_MARK_TYPE_L2_ENCAP],
+								   ecm_classifier_mark_default_get);
+				rcu_assign_pointer(sync_to_ipv4_cb[ECM_CLASSIFIER_MARK_TYPE_L2_ENCAP],
+								   ecm_classifier_mark_default_sync_ipv4);
+				rcu_assign_pointer(sync_to_ipv6_cb[ECM_CLASSIFIER_MARK_TYPE_L2_ENCAP],
+								   ecm_classifier_mark_default_sync_ipv6);
+
+				DEBUG_INFO("MARK: Default callbacks registered\n");
+				// printk(KERN_INFO "MARK: INIT - Default callbacks registered. enabled=%d, deny_mark=0x%x, mode=%d\n", ecm_classifier_mark_enabled, ecm_classifier_mark_deny_value, ecm_classifier_mark_mode);
+				return 0;
 }
 EXPORT_SYMBOL(ecm_classifier_mark_init);
 
@@ -775,12 +773,8 @@
 	ecm_classifier_mark_terminate_pending = true;
 	spin_unlock_bh(&ecm_classifier_mark_lock);
 
-	/*
-	 * Remove the debugfs files recursively.
-	 */
 	if (ecm_classifier_mark_dentry) {
 		debugfs_remove_recursive(ecm_classifier_mark_dentry);
 	}
-
 }
 EXPORT_SYMBOL(ecm_classifier_mark_exit);
